### PR TITLE
Asap Client Authentication Extension. 

### PR DESCRIPTION
**Pull request by jamesmoessis**
**Description:**
Adding an extension which allows the collector to authenticate outgoing requests with ASAP ([Atlassian Service Authentication Protocol](https://s2sauth.bitbucket.io/)). 

It works by implementing `ClientAuthenticator` and setting the authorization header as the encoded JWT. It uses the [go-asap](https://bitbucket.org/atlassian/go-asap/src) library to achieve much of its functionality.

**Link to tracking Issue:**
~~N/A~~  https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/5930

**Testing:** 
We mock key server and a base RoundTripper to achieve validation on signature and fields on the JWT. Configuration and factory are also tested. There are yaml files for good and bad config (one good, multiple bad configurations). 

Additionally, at Atlassian we have been using ASAP in the collector for a long time already.

**Documentation:** 

README.md gives a sample configuration and brief description. Those familiar with ASAP should have little trouble setting it up 

Additionally, @MovieStoreGuy and myself are happy to be the codeowners for this extension.

### CODE REVIEW

**Review by MovieStoreGuy — APPROVED** [review]
I had approved this on our fork of the collector.

**Comment by MovieStoreGuy** [comment]
The related issue is https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/5930

**Comment by MovieStoreGuy** [comment]
Going to collapse my two commits into one in order to rerun the failing tests.

File: go.mod [hunk 1]
```diff
@@ -526,6 +532,8 @@ replace github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elast
 
 replace github.com/open-telemetry/opentelemetry-collector-contrib/exporter/zipkinexporter => ./exporter/zipkinexporter
 
+replace github.com/open-telemetry/opentelemetry-collector-contrib/extension/asapauth => ./extension/asapauth
```
Comment (mx-psi) [inline go.mod:535]: This should be done on a separate PR; read the [contributing doc for new components](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#when-adding-a-new-component)

File: internal/components/components.go [hunk 2]
```diff
@@ -130,6 +131,7 @@ func Components() (component.Factories, error) {
 	delete(factories.Extensions, "pprof")
 
 	extensions := []component.ExtensionFactory{
+		asapauth.NewFactory(),
```
Comment (mx-psi) [inline internal/components/components.go:134]: ditto

File: extension/asapauth/go.mod [hunk 3]
```diff
@@ -0,0 +1,47 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/extension/asapauth
+
+go 1.17
+
+require (
+	bitbucket.org/atlassian/go-asap v0.0.0-20210823164515-3d5971c40c4c
+	github.com/SermoDigital/jose v0.9.2-0.20161205224733-f6df55f235c2
+	github.com/stretchr/testify v1.7.0
+	go.opentelemetry.io/collector v0.38.0
```
Comment (mx-psi) [inline extension/asapauth/go.mod:9]: This should match the version on the `main` branch

File: extension/asapauth/factory.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+		typeStr,
+		createDefaultConfig,
+		createExtension,
+	)
+}
+
+func createExtension(_ context.Context, _ component.ExtensionCreateSettings, cfg config.Extension) (component.Extension, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
```
Comment (mx-psi) [inline extension/asapauth/factory.go:42]: Validation [will be done by the service pipeline during initialization](https://github.com/open-telemetry/opentelemetry-collector/blob/bdcb989da369fd2b06be71966dd6f3259aca7fdc/service/service.go#L46-L48), so there is no need to do this again here.

```suggestion
```

File: extension/asapauth/config.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	TTL int `mapstructure:"ttl_seconds"`
+
+	Issuer string `mapstructure:"issuer"`
+
+	Audience []string `mapstructure:"audience"`
+
+	PrivateKey string `mapstructure:"private_key"`
+}
+
+func (c *Config) Validate() error {
```
Comment (mx-psi) [inline extension/asapauth/config.go:45]: Maybe you want to use `multierr` to return all the errors here instead of the first one?

**Review by mx-psi — CHANGES_REQUESTED** [review]

File: internal/components/components.go (hunk 2 above)
Comment (jamesmoessis) [reply internal/components/components.go]: Removed

File: extension/asapauth/go.mod (hunk 3 above)
Comment (jamesmoessis) [reply extension/asapauth/go.mod]: Fixed!

File: extension/asapauth/config.go (hunk 5 above)
Comment (jamesmoessis) [reply extension/asapauth/config.go:46]: Done!

File: go.mod (hunk 1 above)
Comment (jamesmoessis) [reply go.mod]: Thanks, I've removed all the changes to the root `go.mod`. However, the build fails saying "asapauth/go.mod is not listed in any module set". Since I can't list it in the root go.mod, is there something I can do to make this green? Or is it fine to leave it like this.

Comment (mx-psi) [reply go.mod]: You need to add it on `versions.yaml` (we should add this step to the contributing guidelines, sorry, it's confusing)

**Comment by mx-psi** [comment]
Looks good once added to the `versions.yaml` file!

**Comment by jamesmoessis** [comment]
Thanks @mx-psi, I've updated the `versions.yaml`. I also raised a PR to add it to the contributing guide: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6411

The tests seem pretty flakey right now. I can't seem to get it to go green, and the failures are unrelated to this PR.

**Comment by jamesmoessis** [comment]
Looks like the build passed on my 2nd empty commit 😀

**Review by mx-psi — APPROVED** [review]

**Comment by jamesmoessis** [comment]
Hey @mx-psi, any idea when this will be merged?

**Comment by mx-psi** [comment]
@jamesmoessis I was waiting on @tigrannajaryan to review to mark as `ready-to-merge`, let's give them a couple days since there were some holiday days on the US and after that I will mark this since I see no blockers

**Comment by jamesmoessis** [comment]
Awesome thankyou 😄

**Comment by tigrannajaryan** [comment]
@jpkrohling you are the auth expert, can you please review?

File: extension/asapauth/README.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# ASAP Client Authentication Extension
+
+This extension provides [Atlassian Service Authentication Protocol](https://s2sauth.bitbucket.io/) (ASAP) client 
+credentials for HTTP or RPC based exporters.
```
Comment (jpkrohling) [inline extension/asapauth/README.md:4]:
```suggestion
credentials for HTTP or gRPC based exporters. 
```

File: .github/CODEOWNERS [hunk 7]
```diff
@@ -53,6 +53,7 @@ extension/httpforwarder/                             @open-telemetry/collector-c
 extension/observer/                                  @open-telemetry/collector-contrib-approvers @asuresh4 @jrcamp
 extension/bearertokenauthextension/                  @open-telemetry/collector-contrib-approvers @jpkrohling @pavankrish123
 extension/oauth2clientauthextension/                 @open-telemetry/collector-contrib-approvers @pavankrish123
+extension/asapauth                                   @open-telemetry/collector-contrib-approvers @jamesmoessis @MovieStoreGuy
```
Comment (jpkrohling) [inline .github/CODEOWNERS:56]: Other auth extensions use the suffix `authextension`. I would prefer to have this one follow the same pattern.

File: extension/asapauth/extension.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+func (a AsapClientAuthenticator) Start(_ context.Context, _ component.Host) error {
+	return nil
+}
+
+// Shutdown does nothing and returns nil
+func (a AsapClientAuthenticator) Shutdown(_ context.Context) error {
+	return nil
+}
+
+var _ configauth.ClientAuthenticator = (*AsapClientAuthenticator)(nil)
```
Comment (jpkrohling) [inline extension/asapauth/extension.go:67]: Perhaps move this closer to the type definition?

File: extension/asapauth/extension.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+	provisioner asap.Provisioner
+	privateKey  interface{}
+}
+
+func (a AsapClientAuthenticator) RoundTripper(base http.RoundTripper) (http.RoundTripper, error) {
+	return asap.NewTransportDecorator(a.provisioner, a.privateKey)(base), nil
+}
+
+func (a AsapClientAuthenticator) PerRPCCredentials() (credentials.PerRPCCredentials, error) {
+	token, err := a.provisioner.Provision()
```
Comment (jpkrohling) [inline extension/asapauth/extension.go:41]: IIRC, this is called only once, during the setup of the extension. Your `PerRPCAuth` type would need to handle the caching there, not here.

File: extension/asapauth/README.md [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+```yaml
+extensions:
+  asapclient:
+    key_id: somekeyid
+    issuer: someissuer
+    audience:
+      - someservice
+      - someotherservice
+    private_key: ${ASAP_PRIVATE_KEY}
+    ttl_seconds: 120 # Default: 60
```
Comment (jpkrohling) [inline extension/asapauth/README.md:17]: What does this property do? Will a new token be obtained after the TTL? Would an existing token just be validated after the TTL? If this is explained in another authoritative document, like Asap's documentation, link there.

File: .github/CODEOWNERS (hunk 7 above)
Comment (jamesmoessis) [reply .github/CODEOWNERS]: I've changed the name to `asapauthextension`

File: extension/asapauth/extension.go (hunk 8 above)
Comment (jamesmoessis) [reply extension/asapauth/extension.go]: Done 👍

File: extension/asapauth/extension.go (hunk 9 above)
Comment (jamesmoessis) [reply extension/asapauth/extension.go]: Thanks for pointing this out, you are absolutely right. I have moved the token provisioning logic into `GetRequestMetadata()` in `PerRPCAuth`.

File: extension/asapauth/README.md (hunk 10 above)
Comment (jamesmoessis) [reply extension/asapauth/README.md]: The `exp` (expiration) claim on the JWT is set as `time.Now().Add(TTL)`. So, the token will be cached until it's expired, in which case another token will be provisioned. 

The client caching itself is implementation specific, but is related to what the spec says about token expiry, so I have linked that information here. Additionally, I have added comments to all of these yaml fields with a brief on what they are.

**Comment by jamesmoessis** [comment]
Not sure why the linter is failing: 

```
-package asapauthextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/asapauthextension"
+package asapauthextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/asapauth"
```

Wants me to change the comment back to the old package name

**Comment by mx-psi** [comment]
> Not sure why the linter is failing

You need to install porto and run `make goporto` (see #6498)

File: .github/CODEOWNERS [hunk 11]
```diff
@@ -53,6 +53,7 @@ extension/httpforwarder/                             @open-telemetry/collector-c
 extension/observer/                                  @open-telemetry/collector-contrib-approvers @asuresh4 @jrcamp
 extension/bearertokenauthextension/                  @open-telemetry/collector-contrib-approvers @jpkrohling @pavankrish123
 extension/oauth2clientauthextension/                 @open-telemetry/collector-contrib-approvers @pavankrish123
+extension/asapauthextension                          @open-telemetry/collector-contrib-approvers @jamesmoessis @MovieStoreGuy
```
Comment (jpkrohling) [inline .github/CODEOWNERS:56]: Could you please move it to its correct place when sorted alphabetically? I recently reordered the components on this file, you might need to rebase this PR.

File: extension/asapauthextension/README.md [hunk 12]
```diff
@@ -0,0 +1,36 @@
+# ASAP Client Authentication Extension
+
+This extension provides [Atlassian Service Authentication Protocol](https://s2sauth.bitbucket.io/) (ASAP) client
```
Comment (jpkrohling) [inline extension/asapauthextension/README.md:3]: For my own education, do you have a resource with the differences between it and OAuth 2?

File: extension/asapauthextension/config_test.go [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/config/configtest"
+)
+
+// Test keys. Not for use anywhere but these tests.
+const (
+	PrivateKey = "data:application/pkcs8;kid=test;base64,MIIBUwIBADANBgkqhkiG9w0BAQEFAASCAT0wggE5AgEAAkEA0ZPr5JeyVDoB8RyZqQsx6qUD+9gMFg1/0hgdAvmytWBMXQJYdwkK2dFJwwZcWJVhJGcOJBDfB/8tcbdJd34KZQIDAQABAkBZD20tJTHJDSWKGsdJyNIbjqhUu4jXTkFFPK4Hd6jz3gV3fFvGnaolsD5Bt50dTXAiSCpFNSb9M9GY6XUAAdlBAiEA6MccfdZRfVapxKtAZbjXuAgMvnPtTvkVmwvhWLT5Wy0CIQDmfE8Et/pou0Jl6eM0eniT8/8oRzBWgy9ejDGfj86PGQIgWePqIL4OofRBgu0O5TlINI0HPtTNo12U9lbUIslgMdECICXT2RQpLcvqj+cyD7wZLZj6vrHZnTFVrnyR/cL2UyxhAiBswe/MCcD7T7J4QkNrCG+ceQGypc7LsxlIxQuKh5GWYA=="
```
Comment (jpkrohling) [inline extension/asapauthextension/config_test.go:30]: Does this need to be exported?

File: extension/asapauthextension/extension_test.go [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+	assert.NoError(t, err)
+	assert.NotNil(t, roundTripper)
+
+	req := &http.Request{Method: "Get", Header: map[string][]string{}}
+	resp, err := roundTripper.RoundTrip(req)
+	assert.NoError(t, err)
+	authHeaderValue := resp.Header.Get("Authorization")
+	tokenString := authHeaderValue[7:] // Remove prefix "Bearer "
+
+	validateAsapJwt(t, cfg, tokenString)
```
Comment (jpkrohling) [inline extension/asapauthextension/extension_test.go:72]:
```suggestion
	validateAsapJWT(t, cfg, tokenString)
```

No need to fix this if this is the only thing pending for this PR, although it would be nice :-)

File: extension/asapauthextension/go.mod [hunk 15]
```diff
@@ -0,0 +1,46 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/extension/asapauth
```
Comment (jpkrohling) [inline extension/asapauthextension/go.mod:1]: This probably needs to have the `extension` suffix, which is likely the reason the lint check is failing.

```suggestion
module github.com/open-telemetry/opentelemetry-collector-contrib/extension/asapauthextension
```

**Comment by jamesmoessis** [comment]
Something has gone... very wrong with this branch, I apoplogise for that. Currently fixing it up.

**Comment by jamesmoessis** [comment]
I've messed up this branch, will raise a new PR. Sorry for the inconvenience
